packages/auth/src/session.ts

```diff
@@ -67,109 +90,192 @@ export class Session implements SessionDump {
     const network = await this.getNetwork(net)
+
     const url = network.sequenceApiUrl
+    if (!url) throw Error('No chaind url')
 
+    const jwtAuth = (await this.getJWT(network, tryAuth)).token
+
+    return new ArcadeumAPIClient(url, jwtAuth, async () => {
+      this.jwts.delete(url)
+
+      const proofString = this.proofStrings.get(url)
+      if (proofString && !await this.isProofStringValid(await proofString.proofString, network)) {
+        this.proofStrings.delete(url)
+      }
+
+      return (await this.getJWT(network, true)).token
+    })
+  }
+
+  private async getJWT(network: NetworkConfig, tryAuth: boolean): Promise<SessionJWT> {
+    const url = network.sequenceApiUrl
     if (!url) throw Error('No chaind url')
 
-    const jwt = this.jwts[url]
+    // check if we already have or are waiting for a token
+    if (this.jwts.has(url)) {
+      const jwt = this.jwts.get(url)!
+
+      const token = await jwt.token
+
+      if (this.now() < jwt.expiration) {
+        return { token, expiration: jwt.expiration }
+      }
 
-    if (!jwt || jwt.expiration < this.now()) {
-      if (tryAuth) return this.auth(net)
-      throw Error('Not authenticated')
+      // token expired, delete it and get a new one
+      this.jwts.delete(url)
     }
 
-    const api = new ArcadeumAPIClient(url)
-    api.jwtAuth = jwt.token
-    return api
+    if (!tryAuth) {
+      throw new Error('no auth token in memory')
+    }
+
+    const { proofString, expiration } = this.getProofString(url)
+
+    const jwt = {
+      token: proofString.then(async proofString => {
+        const api = new ArcadeumAPIClient(url)
+
+        const authResp = await api.getAuthToken({ ewtString: proofString })
+
+        if (authResp?.status === true && authResp.jwtToken.length !== 0) {
+          return authResp.jwtToken
+        } else {
+          if (!await this.isProofStringValid(proofString, network)) {
+            this.proofStrings.delete(url)
+          }
+          throw new Error('no auth token from server')
+        }
+      }).catch(reason => {
+        this.jwts.delete(url)
+        throw reason
+      }),
+      expiration
+    }
+    this.jwts.set(url, jwt)
+
+    jwt.token.then(() => {
+      this.onAuthCallbacks.forEach(cb => { try { cb({ status: 'fulfilled', value: undefined }) } catch {} })
+    }).catch((reason: any) => {
+      this.onAuthCallbacks.forEach(cb => { try { cb({ status: 'rejected', reason }) } catch {} })
+    })
+
+    const token = await jwt.token
+    return { token, expiration }
   }
 
-  async performAuthRequest(net: NetworkConfig | number): Promise<void> {
-    const network = this.getNetwork(net)
+  private getProofString(url: string): ProofStringPromise {
+    // check if we already have or are waiting for a proof string
+    if (this.proofStrings.has(url)) {
+      const proofString = this.proofStrings.get(url)!
+
+      if (this.now() < proofString.expiration) {
```

Comment:
What happens if our clock is out of sync with the server, so the server will reject our proof string, but it looks good to us?

Reply:
Pretty much all `GetAuthToken` requests to the API will fail in that case, but that's no different from the current behaviour.

Reply:
Is there something clever we can do about checking our assumed system time vs the timestamp returned by some trusted service to avoid the GetAuthToken spam in these cases?